python/hail/methods/statgen.py

```diff
@@ -2158,64 +2147,61 @@ def genetic_relatedness_matrix(dataset) -> KinshipMatrix:
 
         G_{ik} = \\frac{1}{m} \\sum_{j=1}^m \\frac{(C_{ij}-2p_j)(C_{kj}-2p_j)}{2 p_j (1-p_j)}
 
-    Warning
-    -------
-    Since Hardy-Weinberg normalization cannot be applied to variants that
-    contain only reference alleles or only alternate alleles, all such variants
-    are removed prior to calcularing the GRM.
+    Note that variants for which the alternate allele frequency is zero or one are not
+    normalizable, and therefore removed prior to calculating the GRM.
 
     Parameters
     ----------
-    dataset : :class:`.MatrixTable`
-        Dataset to sample from.
+    call_expr : :class:`.CallExpression`
+        Entry-indexed call expression.
 
     Returns
     -------
     :class:`.genetics.KinshipMatrix`
-        Genetic Relatedness Matrix for all samples.
+        Genetic relatedness matrix for all samples.
     """
-    require_col_key_str(dataset, 'genetic_relatedness_matrix')
-    dataset = require_biallelic(dataset, "genetic_relatedness_matrix")
-    dataset = dataset.annotate_rows(AC=agg.sum(dataset.GT.n_alt_alleles()),
-                                    n_called=agg.count_where(hl.is_defined(dataset.GT)))
-    dataset = dataset.filter_rows((dataset.AC > 0) & (dataset.AC < 2 * dataset.n_called)).persist()
+    mt = matrix_table_source('genetic_relatedness_matrix/call_expr', call_expr)
+    require_col_key_str(mt, 'genetic_relatedness_matrix')
+
+    col_keys = mt.cols().select(*mt.col_key)
 
-    n_variants = dataset.count_rows()
+    mt = mt.select_entries(__gt=call_expr.n_alt_alleles())
+    mt = mt.annotate_rows(__AC=agg.sum(mt.__gt),
+                          __n_called=agg.count_where(hl.is_defined(mt.__gt)))
+    mt = mt.filter_rows((mt.__AC > 0) & (mt.__AC < 2 * mt.__n_called)).persist()
+
+    n_variants = mt.count_rows()
     if n_variants == 0:
         raise FatalError("Cannot run GRM: found 0 variants after filtering out monomorphic sites.")
     info("Computing GRM using {} variants.".format(n_variants))
 
-    normalized_genotype_expr = hl.bind(
-        dataset.AC / dataset.n_called,
-        lambda mean_gt: hl.cond(hl.is_defined(dataset.GT),
-                                (dataset.GT.n_alt_alleles() - mean_gt) /
-                                hl.sqrt(mean_gt * (2 - mean_gt) * n_variants / 2),
-                                0))
+    mt = mt.annotate_rows(__mean_gt=mt.__AC / mt.__n_called)
+    mt = mt.annotate_rows(
+        __hwe_std_dev=hl.sqrt(mt.__mean_gt * (2 - mt.__mean_gt) * n_variants / 2))
+
+    normalized_gt = hl.or_else((mt.__gt - mt.__mean_gt) / mt.__hwe_std_dev, 0.0)
 
-    bm = BlockMatrix.from_entry_expr(normalized_genotype_expr)
-    dataset.unpersist()
+    bm = BlockMatrix.from_entry_expr(normalized_gt)
+    mt.unpersist()
```

Comment:
I don't see a persist  call in this method. Should this be removed?

Reply:
It's there in GRM and RRM, but I'll move persist to it's own line so easier to spot.